guides/api-guideline.md

```diff
@@ -0,0 +1,131 @@
+| `bool`          | A boolean value                                               |
+| `bytes`         | A sequence of bytes                                           |
+| `list<TYPE>`    | A list of elements of type TYPE                               |
+| `set<TYPE>`     | A set of elements of type TYPE                                |
+| `map<TYPE>`     | An immutable map of elements of type TYPE                     |
+| `date`          | A date value (ISO 8601 calendar date)                         |
+| `time`          | A time value without date or timezone (nanosecond precision)  |
+| `dateTime`      | A date and time value without timezone (nanosecond precision) |
+| `zonedDateTime` | A date and time value with timezone (nanosecond precision)    |
+
+### Enumerations
+
+Enumerations can be defined using the following syntax:
+```
+enum EnumName {
+    VALUE1
+    VALUE2
+}
+```
+
+### Attribute annotations
+
+Attribute annotations can be used to provide additional information about attributes in complex data types.
+The following annotations should be used:
+- `@immutable`: Indicates that the field is immutable and cannot be changed after creation.
+- `@optional`: Indicates that the field is optional and can be null or undefined.
+- `@default(value)`: Indicates that the field has a default value.
+- `@min(value)`: Indicates the minimum value for numeric fields.
+- `@max(value)`: Indicates the maximum value for numeric fields.
+- `@minLength(value)`: Indicates the minimum length for string fields.
+- `@maxLength(value)`: Indicates the maximum length for string fields.
+- `@pattern(regex)`: Indicates a regex pattern that the string field must match.
+
+### Method annotations
+
+Method annotations can be used to provide additional information about methods.
+The following annotations should be used:
+- `@async`: Indicates that the method is asynchronous and returns a promise or future.
+- `@throws(error-type)`: Indicates that the method can throw an exception/error.
+
+### Method definitions
+
+Methods can be defined using the following syntax:
+
+```
+ReturnType methodName(param1: DataType1, param2: DataType2)
+```
+
+The following attribute annotations can be used on method parameters: `@optional`, `@min(value)`, `@max(value)`, `@minLength(value)`, `@maxLength(value)`, `@pattern(regex)`
+
+The following attribute annotations can be used on method return types: `@immutable`, `@optional`
+
+A complete method definition example:
+```
+@async
+@throws(parse-error)
+@optional ResponseType fetchData(id: uint64, @optional filter: string)
+```
+
+### Complex data types
+
+Complex data types can be defined using the basic data types and other complex data types.
+The following syntax should be used to define complex data types:
+```
+DataTypeName {
+    fieldName1: DataType1
+    fieldName2: DataType2
+    ReturnType methodName(param1: DataType1, param2: DataType2)
+}
+```
```

Comment:
Should `public` and `private` be specified? Or should everything defined be assumed to be `public` since anything but is just implementation details?

Reply:
Everything that is defined must be public. We should not define global private methods